### PR TITLE
Action mailchimp new actions

### PR DESCRIPTION
**Pull request by bush4u**

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible

File: components/mailchimp/actions/delete-a-list/delete-a-list.mjs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    mailchimp,
+    listId: {
+      type: "string",
+      label: "List id",
+      description: "The unique ID for the list.",
+    },
+  },
+  async run({ $ }) {
+    const response = await this.mailchimp.updateList($, this.listId);
+    response && $.export("$summary", "List deleted");
```
Comment (lcaresia) [inline components/mailchimp/actions/delete-a-list/delete-a-list.mjs:19]: You can add `Successfully` for the other too.

```suggestion
    response && $.export("$summary", "Successfully deleted list");
```

File: components/mailchimp/actions/edit-campaign-template-content/edit-campaign-template-content.mjs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        label: "Archive content",
+        description: "The base64-encoded representation of the archive file",
+      };
+    }
+    return props;
+  },
+  async run({ $ }) {
+    let variateContents;
+    try {
+      variateContents = this.variate_contents?.map((content) => JSON.parse(content));
```
Comment (lcaresia) [inline components/mailchimp/actions/edit-campaign-template-content/edit-campaign-template-content.mjs:88]: Maybe here need another `empty string` validation?

File: components/mailchimp/actions/get-a-campaign/get-a-campaign.mjs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      fields,
+      excludeFields,
+      campaignId,
+    } = this;
+    const payload =  removeNullEntries({
+      fields: fields.join(","),
+      exclude_fields: excludeFields.join(","),
+      campaignId,
+    });
+    const response = await this.mailchimp.getACampaign($, payload);
```
Comment (lcaresia) [inline components/mailchimp/actions/get-a-campaign/get-a-campaign.mjs:40]:
```suggestion
    const response = await this.mailchimp.getCampaign($, payload);
```

File: components/mailchimp/mailchimp.app.mjs [hunk 4]
```diff
@@ -583,5 +616,172 @@ export default {
...
       return await this._withRetries(() => this.api().reports.getCampaignOpenDetails(campaignId));
     },
+    async searchCampaign($, params) {
+      return this._makeRequest({
+        $,
+        params,
+        path: "/search-campaigns",
+      });
+    },
+    async getACampaign($, {
```
Comment (lcaresia) [inline components/mailchimp/mailchimp.app.mjs:626]:
```suggestion
    async getCampaign($, {
```

File: components/mailchimp/actions/get-a-campaign/get-a-campaign.mjs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    response && $.export("$summary", "Campaign found");
```
Comment (lcaresia) [inline components/mailchimp/actions/get-a-campaign/get-a-campaign.mjs:41]: Use `retrieved` isntead of `found`

```suggestion
    response && $.export("$summary", "Successfully retrieved campaign");
```

File: components/mailchimp/actions/get-a-list/get-a-list.mjs [hunk 6]
```diff
@@ -0,0 +1,46 @@
+import { removeNullEntries } from "../../common/utils.mjs";
+import mailchimp from "../../mailchimp.app.mjs";
+
+export default {
+  key: "mailchimp-get-a-list",
+  name: "Get a List",
```
Comment (lcaresia) [inline components/mailchimp/actions/get-a-list/get-a-list.mjs:6]: Remove the `a` between the mainly words. 
You can do this for the other too.

```suggestion
  name: "Get List",
```

File: components/mailchimp/actions/get-a-list/get-a-list.mjs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+      optional: true,
+    },
+    excludeFields: {
+      type: "string[]",
+      label: "Exclude Fields",
+      description: "A string list of fields to exclude_fields. Reference parameters of sub-objects with dot notation.",
+    },
+    listId: {
+      type: "string",
+      label: "List id",
```
Comment (lcaresia) [inline components/mailchimp/actions/get-a-list/get-a-list.mjs:25]: Normally we use the `id` word in uppercase on label.

You can do this for the other too.

```suggestion
      label: "List ID",
```

File: components/mailchimp/actions/get-a-list/get-a-list.mjs [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+      description: "The unique ID for the list.",
+    },
+    includeTotalContacts: {
+      type: "boolean",
+      label: "Include total contacts?",
```
Comment (lcaresia) [inline components/mailchimp/actions/get-a-list/get-a-list.mjs:30]:
```suggestion
      label: "Include total contacts",
```

File: components/mailchimp/actions/get-a-list/get-a-list.mjs [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+      description: "Return the total_contacts field in the stats response, which contains an approximate count of all contacts in any state.",
+      optional: true,
+    },
+  },
+  async run({ $ }) {
+    const payload = removeNullEntries({
+      listId: this.listId,
+      fields: this.fields.join(","),
+      exclude_fields: this.excludeFields.join(","),
```
Comment (lcaresia) [inline components/mailchimp/actions/get-a-list/get-a-list.mjs:39]: The same of above.

File: components/mailchimp/actions/get-campaign-report/get-campaign-report.mjs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  key: "mailchimp-get-a-campaign-report",
+  name: "Get a Campaign Report",
+  description: "Gets a campaign report. [See docs here](https://mailchimp.com/developer/marketing/api/campaign-advice/)",
+  version: "0.0.1",
+  type: "action",
+  props: {
+    mailchimp,
+    campaignId: {
+      type: "string",
+      label: "Campaign Id",
```
Comment (lcaresia) [inline components/mailchimp/actions/get-campaign-report/get-campaign-report.mjs:14]:
```suggestion
      label: "Campaign ID",
```

File: components/mailchimp/actions/get-campaign-report/get-campaign-report.mjs [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+      description: "A string list of fields to exclude_fields. Reference parameters of sub-objects with dot notation.",
+    },
+  },
+  async run({ $ }) {
+    const {
+      fields,
+      excludeFields,
+      campaignId,
+    } = this;
+    const payload =  removeNullEntries({
```
Comment (lcaresia) [inline components/mailchimp/actions/get-campaign-report/get-campaign-report.mjs:35]:
```suggestion
    const payload = removeNullEntries({
```

File: components/mailchimp/actions/get-list-activities/get-list-activities.mjs [hunk 12]
```diff
@@ -0,0 +1,50 @@
+import { removeNullEntries } from "../../common/utils.mjs";
+import mailchimp from "../../mailchimp.app.mjs";
+
+export default {
+  key: "mailchimp-get-a-list",
+  name: "List List's Activities",
```
Comment (lcaresia) [inline components/mailchimp/actions/get-list-activities/get-list-activities.mjs:6]:
```suggestion
  name: "Get List Activities",
```

File: components/mailchimp/actions/get-list-member-activity/get-list-member-activity.mjs [hunk 13]
```diff
@@ -0,0 +1,56 @@
+import { removeNullEntries } from "../../common/utils.mjs";
+import mailchimp from "../../mailchimp.app.mjs";
+
+export default {
+  key: "mailchimp-get-a-list",
+  name: "List List Members' Activities",
```
Comment (lcaresia) [inline components/mailchimp/actions/get-list-member-activity/get-list-member-activity.mjs:6]:
```suggestion
  name: "Get List Members Activities",
```

File: components/mailchimp/actions/get-list-member-tags/get-list-member-tags.mjs [hunk 14]
```diff
@@ -0,0 +1,57 @@
+import { removeNullEntries } from "../../common/utils.mjs";
+import mailchimp from "../../mailchimp.app.mjs";
+
+export default {
+  key: "mailchimp-get-list-member-tags",
+  name: "Get List Members' Tags",
```
Comment (lcaresia) [inline components/mailchimp/actions/get-list-member-tags/get-list-member-tags.mjs:6]:
```suggestion
  name: "Get List Members Tags",
```

File: components/mailchimp/actions/search-lists/search-lists.mjs [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+      before_campaign_last_sent: this.beforeCampaignLastSent,
+      since_campaign_last_sent: this.sinceCampaignLastSent,
+      email: this.email,
+      sort_field: this.sortField,
+      sort_dir: this.sortDir,
+      has_ecommerce_store: this.hasEcommerceStore,
+      include_total_contacts: this.includeTotalContacts,
+    });
+    const response = await this.mailchimp.searchLists($, payload);
+    // response?.lists?.length > 0 && $.export("$summary", "Lists found");
```
Comment (lcaresia) [inline components/mailchimp/actions/search-lists/search-lists.mjs:109]:
```suggestion
```

File: components/mailchimp/actions/search-lists/search-lists.mjs [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+    console.log(response);
```
Comment (lcaresia) [inline components/mailchimp/actions/search-lists/search-lists.mjs:110]:
```suggestion
```

File: components/mailchimp/actions/update-campaign/update-campaign.mjs [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+    },
+    segmentMatch: {
+      type: "string",
+      label: "Segment match",
+      description: "Segment match type.",
+      optional: true,
+      options: [
+        "any",
+        "all",
+      ],
```
Comment (lcaresia) [inline components/mailchimp/actions/update-campaign/update-campaign.mjs:43]: You can move this for a `common/constants.mjs` file.

```suggestion
      options: constants.SEGMENT_MATCHES,
```

File: components/mailchimp/common/constants.mjs [hunk 18]
```diff
@@ -0,0 +1,9 @@
+const HTTP_PROTOCOL = "https://";
+const BASE_URL = "api.mailchimp.com";
+const VERSION_PATH = "/3.0";
+
+export default {
+  HTTP_PROTOCOL,
+  BASE_URL,
+  VERSION_PATH,
+};
```
Comment (lcaresia) [inline components/mailchimp/common/constants.mjs:9]:
```suggestion
export default {
  HTTP_PROTOCOL: "https://",
  BASE_URL: "api.mailchimp.com",
  VERSION_PATH: "/3.0",
  SEGMENT_MATCHES: [
    "any",
    "all",
  ],
};

```

File: components/mailchimp/mailchimp.app.mjs [hunk 19]
```diff
@@ -129,6 +131,37 @@ export default {
     _server() {
       return this.$auth.dc;
     },
+    _getHeader() {
```
Comment (lcaresia) [inline components/mailchimp/mailchimp.app.mjs:134]:
```suggestion
    _getHeaders() {
```

File: components/mailchimp/mailchimp.app.mjs [hunk 20]
```diff
@@ -129,6 +131,37 @@ export default {
...
+        $,
+        method = "get",
+        path,
+        params,
+        data,
+      } = args;
+      const config = {
+        method,
+        url: this._getUrl(path),
+        headers: this._getHeader(),
```
Comment (lcaresia) [inline components/mailchimp/mailchimp.app.mjs:159]:
```suggestion
        headers: this._getHeaders(),
```

**Review by lcaresia — CHANGES_REQUESTED** [review]
Everything looks good, just need a few updates.

Some points you can improve the component:

- [Try to define props inside the app and reuse them with propDefinition;](https://pipedream.com/docs/components/api/#props)
- [Implement async options whenever possible;](https://pipedream.com/docs/components/migrating/#advanced-using-async-options)
- Keep the number of props small, don't need put all params you found in the API Docs as props, you can choose the 3~5 most importants.

File: components/mailchimp/actions/add-segment-member/add-segment-member.mjs [hunk 21]
```diff
@@ -0,0 +1,37 @@
+import mailchimp from "../../mailchimp.app.mjs";
+
+export default {
+  key: "mailchimp-add-segment-member",
+  name: "Add A Member To A Segment",
```
Comment (lcaresia) [inline components/mailchimp/actions/add-segment-member/add-segment-member.mjs:5]:
```suggestion
  name: "Add Member To Segment",
```

File: components/mailchimp/actions/add-subscriber-to-tag/add-subscriber-to-tag.mjs [hunk 22]
```diff
@@ -1,56 +1,48 @@
...
-import { axios } from "@pipedream/platform";
+import constants from "../../common/constants.mjs";
+import {
+  formatArrayStrings, removeNullEntries,
+} from "../../common/utils.mjs";
+import mailchimp from "../../mailchimp.app.mjs";
 
 export default {
   key: "mailchimp-add-subscriber-to-tag",
   name: "Add Subscriber to Tag",
```
Comment (lcaresia) [inline components/mailchimp/actions/add-subscriber-to-tag/add-subscriber-to-tag.mjs:10]:
```suggestion
  name: "Add Subscriber To Tag",
```

File: components/mailchimp/actions/create-campaign/create-campaign.mjs [hunk 23]
```diff
@@ -23,21 +22,25 @@ export default {
...
+      label: "Saved segment id",
       type: "integer",
       description: "The id for an existing saved segment.",
       optional: true,
     },
-    prebuilt_segment_id: {
+    prebuiltSegmentId: {
+      label: "Prebuilt segment id",
       type: "string",
       description: "The prebuilt segment id, if a prebuilt segment has been designated for this campaign.",
```
Comment (lcaresia) [inline components/mailchimp/actions/create-campaign/create-campaign.mjs:39]:
```suggestion
      description: "The prebuilt segment ID, if a prebuilt segment has been designated for this campaign.",
```

File: components/mailchimp/actions/create-campaign/create-campaign.mjs [hunk 24]
```diff
@@ -46,147 +49,176 @@ export default {
...
+      label: "Use conversation",
       type: "boolean",
       description: "Use Mailchimp Conversation feature to manage out-of-office replies.",
       optional: true,
     },
-    to_name: {
+    toName: {
+      label: "To name",
       type: "string",
       description: "The campaign's custom 'To' name.",
```
Comment (lcaresia) [inline components/mailchimp/actions/create-campaign/create-campaign.mjs:97]: Why this `To` is between single quotes?

File: components/mailchimp/actions/create-campaign/create-campaign.mjs [hunk 25]
```diff
@@ -46,147 +49,176 @@ export default {
...
       optional: true,
     },
-    folder_id: {
+    folderId: {
+      label: "Folder ID",
       type: "string",
       description: "If the campaign is listed in a folder, the id for that folder.",
```
Comment (lcaresia) [inline components/mailchimp/actions/create-campaign/create-campaign.mjs:103]:
```suggestion
      description: "If the campaign is listed in a folder, the ID for that folder.",
```

File: components/mailchimp/actions/create-campaign/create-campaign.mjs [hunk 26]
```diff
@@ -46,147 +49,176 @@ export default {
...
+      label: "FB comments",
       type: "boolean",
       description: "Allows Facebook comments on the campaign (also force-enables the Campaign Archive toolbar). Defaults to true.",
       optional: true,
     },
-    template_id: {
+    templateId: {
+      label: "Template ID",
       type: "integer",
       description: "The id of the template to use.",
```
Comment (lcaresia) [inline components/mailchimp/actions/create-campaign/create-campaign.mjs:145]:
```suggestion
      description: "The ID of the template to use.",
```

File: components/mailchimp/actions/delete-campaign/delete-campaign.mjs [hunk 27]
```diff
@@ -0,0 +1,23 @@
...
+  name: "Delete Campaign",
+  description: "Delete a specific campaign. [See docs here](https://mailchimp.com/developer/marketing/api/campaigns/delete-campaign/)",
+  version: "0.0.1",
+  type: "action",
+  props: {
+    mailchimp,
+    campaignId: {
+      type: "string",
+      label: "Campaign ID",
+      description: "The unique id for the campaign",
```
Comment (lcaresia) [inline components/mailchimp/actions/delete-campaign/delete-campaign.mjs:14]:
```suggestion
      description: "The unique ID for the campaign",
```

File: components/mailchimp/actions/delete-list-member/delete-list-member.mjs [hunk 28]
```diff
@@ -0,0 +1,31 @@
+import mailchimp from "../../mailchimp.app.mjs";
+
+export default {
+  key: "mailchimp-delete-list-member",
+  name: "Delete A List Member",
```
Comment (lcaresia) [inline components/mailchimp/actions/delete-list-member/delete-list-member.mjs:5]:
```suggestion
  name: "Delete List Member",
```

File: components/mailchimp/actions/delete-list-member/delete-list-member.mjs [hunk 29]
```diff
@@ -0,0 +1,31 @@
...
+  description: "Permanently deletes a member. [See docs here](https://mailchimp.com/developer/marketing/api/list-members/delete-list-member/)",
+  version: "0.0.1",
+  type: "action",
+  props: {
+    mailchimp,
+    listId: {
+      type: "string",
+      label: "List ID",
+      description: "The unique id for the list.",
```
Comment (lcaresia) [inline components/mailchimp/actions/delete-list-member/delete-list-member.mjs:14]:
```suggestion
      description: "The unique ID for the list.",
```

File: components/mailchimp/actions/add-remove-member-tags/add-remove-member-tags.mjs [hunk 30]
```diff
@@ -0,0 +1,46 @@
...
+    },
+    subscriberHash: {
+      type: "string",
+      label: "Subscriber hash",
+      description: "The MD5 hash of the lowercase version of the list member's email address. This endpoint also accepts a list member's email address or contact_id.",
+    },
+    tags: {
+      type: "string[]",
+      label: "Tags",
+      description: `Stringified object list of fields to return. name, or status (Possible status values: "inactive" or "active")  properties allowed.
```
Comment (lcaresia) [inline components/mailchimp/actions/add-remove-member-tags/add-remove-member-tags.mjs:28]:
```suggestion
      description: `Stringified object list of fields to return. name, or status (Possible status values: "inactive" or "active") properties allowed.
```

File: components/mailchimp/actions/add-remove-member-tags/add-remove-member-tags.mjs [hunk 31]
```diff
@@ -0,0 +1,46 @@
...
+        }\``,
+    },
+  },
+  async run({ $ }) {
+    const payload = removeNullEntries({
+      listId: this.listId,
+      subscriberHash: this.subscriberHash,
+      tags: formatArrayStrings(this.tags, constants.ALLOWED_TAG_KEYS, "Tags"),
+    });
+    const response =  await this.mailchimp.addRemoveListMemberTags($, payload);
```
Comment (lcaresia) [inline components/mailchimp/actions/add-remove-member-tags/add-remove-member-tags.mjs:42]:
```suggestion
    const response = await this.mailchimp.addRemoveListMemberTags($, payload);
```

File: components/mailchimp/actions/add-segment-member/add-segment-member.mjs [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
+      description: "Email address for a subscriber.",
+    },
+  },
+  async run({ $ }) {
+    const payload = {
+      email_address: this.emailAddress,
+      segmentId: this.segmentId,
+      listId: this.listId,
+    };
+    const response =  await this.mailchimp.addSegmentMember($, payload);
```
Comment (lcaresia) [inline components/mailchimp/actions/add-segment-member/add-segment-member.mjs:33]:
```suggestion
    const response = await this.mailchimp.addSegmentMember($, payload);
```

File: components/mailchimp/actions/add-subscriber-to-tag/add-subscriber-to-tag.mjs [hunk 33]
```diff
@@ -1,56 +1,48 @@
...
-      type: "string",
-      description: "The status for the tag on the member, pass in active to add a tag or inactive to remove it.",
-      options: [
-        "active",
-        "inactive",
-      ],
+    tags: {
+      type: "string[]",
+      label: "Tags",
+      description: `Stringified object list of fields to return. name, or status (Possible status values: "inactive" or "active")  properties allowed.
```
Comment (lcaresia) [inline components/mailchimp/actions/add-subscriber-to-tag/add-subscriber-to-tag.mjs:29]:
```suggestion
      description: `Stringified object list of fields to return. name, or status (Possible status values: "inactive" or "active") properties allowed.
```

File: components/mailchimp/actions/delete-campaign/delete-campaign.mjs [hunk 34]
```diff
@@ -0,0 +1,23 @@
...
+    },
+  },
+  async run({ $ }) {
+    const {  campaignId } = this;
```
Comment (lcaresia) [inline components/mailchimp/actions/delete-campaign/delete-campaign.mjs:18]:
```suggestion
    const { campaignId } = this;
```

File: components/mailchimp/actions/get-campaign/get-campaign.mjs [hunk 35]
```diff
@@ -0,0 +1,45 @@
...
+      ],
+    },
+  },
+  async run({ $ }) {
+    const {
+      fields,
+      excludeFields,
+      campaignId,
+    } = this;
+    const payload =  removeNullEntries({
```
Comment (lcaresia) [inline components/mailchimp/actions/get-campaign/get-campaign.mjs:36]:
```suggestion
    const payload = removeNullEntries({
```

File: components/mailchimp/actions/remove-segment-member/remove-segment-member.mjs [hunk 36]
```diff
@@ -0,0 +1,37 @@
...
+      description: "The MD5 hash of the lowercase version of the list member's email address.",
+    },
+  },
+  async run({ $ }) {
+    const payload = {
+      subscriberHash: this.subscriberHash,
+      segmentId: this.segmentId,
+      listId: this.listId,
+    };
+    const response =  await this.mailchimp.removeSegmentMember($, payload);
```
Comment (lcaresia) [inline components/mailchimp/actions/remove-segment-member/remove-segment-member.mjs:33]:
```suggestion
    const response = await this.mailchimp.removeSegmentMember($, payload);
```

File: components/mailchimp/mailchimp.app.mjs [hunk 37]
```diff
@@ -129,6 +151,49 @@ export default {
...
+      return `/campaigns/${campaignId || ""}`;
+    },
+    _listPath(listId = null) {
+      return `/lists/${listId || ""}`;
+    },
+    _segmentMemberPath(listId, segmentId, subscriberHash = null) {
+      return  `${this._listPath(listId)}/segments/${segmentId}/members/${subscriberHash || ""}`;
+    },
+    _listMemberPath(listId,  subscriberHash) {
+      return  `${this._listPath(listId)}/members/${subscriberHash}`;
```
Comment (lcaresia) [inline components/mailchimp/mailchimp.app.mjs:178]:
```suggestion
    _listMemberPath(listId, subscriberHash) {
      return `${this._listPath(listId)}/members/${subscriberHash}`;
```

File: components/mailchimp/actions/add-note-to-subscriber/add-note-to-subscriber.mjs [hunk 38]
```diff
@@ -1,43 +1,38 @@
 // legacy_hash_id: a_wdijlV
```
Comment (lcaresia) [inline components/mailchimp/actions/add-note-to-subscriber/add-note-to-subscriber.mjs:1]:
```suggestion
```

File: components/mailchimp/actions/add-or-update-subscriber/add-or-update-subscriber.mjs [hunk 39]
```diff
@@ -1,35 +1,38 @@
 // legacy_hash_id: a_RAiaJ1
```
Comment (lcaresia) [inline components/mailchimp/actions/add-or-update-subscriber/add-or-update-subscriber.mjs:1]:
```suggestion
```

File: components/mailchimp/actions/add-or-update-subscriber/add-or-update-subscriber.mjs [hunk 40]
```diff
@@ -61,102 +66,115 @@ export default {
...
+      label: "Longitude",
       type: "integer",
       description: "The location longitude.",
       optional: true,
     },
-    marketing_permission_id: {
+    marketingPermissionId: {
+      label: "Marketing permission id",
       type: "string",
       description: "The id for the marketing permission on the list.",
```
Comment (lcaresia) [inline components/mailchimp/actions/add-or-update-subscriber/add-or-update-subscriber.mjs:108]:
```suggestion
      label: "Marketing permission ID",
      type: "string",
      description: "The ID for the marketing permission on the list.",
```

File: components/mailchimp/actions/add-remove-member-tags/add-remove-member-tags.mjs [hunk 41]
```diff
@@ -0,0 +1,46 @@
...
+  name: "Add Or Remove Members Tags",
+  description: "Add or remove member tags. [See docs here](https://mailchimp.com/developer/marketing/api/list-member-tags/add-or-remove-member-tags/)",
+  version: "0.0.1",
+  type: "action",
+  props: {
+    mailchimp,
+    listId: {
+      type: "string",
+      label: "List ID",
+      description: "The unique id for the list.",
```
Comment (lcaresia) [inline components/mailchimp/actions/add-remove-member-tags/add-remove-member-tags.mjs:18]:
```suggestion
      description: "The unique ID for the list.",
```

File: components/mailchimp/actions/add-segment-member/add-segment-member.mjs [hunk 42]
```diff
@@ -0,0 +1,37 @@
...
+    mailchimp,
+    listId: {
+      type: "string",
+      label: "List ID",
+      description: "The unique ID for the list.",
+    },
+    segmentId: {
+      type: "string",
+      label: "Segment id",
+      description: "The unique id for the segment.",
```
Comment (lcaresia) [inline components/mailchimp/actions/add-segment-member/add-segment-member.mjs:19]:
```suggestion
      label: "Segment ID",
      description: "The unique ID for the segment.",
```

File: components/mailchimp/actions/create-campaign/create-campaign.mjs [hunk 43]
```diff
@@ -46,147 +49,176 @@ export default {
...
+      label: "Auto tweet",
       type: "boolean",
       description: "Automatically tweet a link to the campaign archive page when the campaign is sent.",
       optional: true,
     },
-    auto_fb_post: {
+    autoFbPost: {
+      label: "Auto fb post",
       type: "any",
       description: "An array of Facebook page ids to auto-post to.",
```
Comment (lcaresia) [inline components/mailchimp/actions/create-campaign/create-campaign.mjs:133]:
```suggestion
      description: "An array of Facebook page ID's to auto-post to.",
```

File: components/mailchimp/actions/edit-campaign-template-content/edit-campaign-template-content.mjs [hunk 44]
```diff
@@ -0,0 +1,109 @@
...
+  name: "Edit A Campaign Template Content",
+  description: "Edits a defined content area of a custom HTML template. [See docs here](https://mailchimp.com/developer/marketing/api/campaign-content/set-campaign-content/)",
+  version: "0.0.1",
+  type: "action",
+  props: {
+    mailchimp,
+    campaignId: {
+      type: "string",
+      label: "Campaign ID",
+      description: "The unique id for the campaign",
```
Comment (lcaresia) [inline components/mailchimp/actions/edit-campaign-template-content/edit-campaign-template-content.mjs:18]:
```suggestion
      description: "The unique ID for the campaign",
```

File: components/mailchimp/actions/get-list-member-activity/get-list-member-activity.mjs [hunk 45]
```diff
@@ -0,0 +1,55 @@
...
+  name: "Get List Member Activities",
+  description: "Get the last 50 events of a member's activity on a specific list. [See docs here](https://mailchimp.com/developer/marketing/api/list-activity/view-recent-activity-50/)",
+  version: "0.0.1",
+  type: "action",
+  props: {
+    mailchimp,
+    listId: {
+      type: "string",
+      label: "List ID",
+      description: "The unique id for the list.",
```
Comment (lcaresia) [inline components/mailchimp/actions/get-list-member-activity/get-list-member-activity.mjs:17]:
```suggestion
      description: "The unique ID for the list.",
```

File: components/mailchimp/actions/get-list-member-tags/get-list-member-tags.mjs [hunk 46]
```diff
@@ -0,0 +1,53 @@
...
+  name: "Get List Member Tags",
+  description: "Retrieves a list of all member tags. [See docs here](https://mailchimp.com/developer/marketing/api/list-member-tags/list-member-tags/)",
+  version: "0.0.1",
+  type: "action",
+  props: {
+    mailchimp,
+    listId: {
+      type: "string",
+      label: "List ID",
+      description: "The unique id for the list.",
```
Comment (lcaresia) [inline components/mailchimp/actions/get-list-member-tags/get-list-member-tags.mjs:15]:
```suggestion
      description: "The unique ID for the list.",
```

File: components/mailchimp/actions/update-campaign/update-campaign.mjs [hunk 47]
```diff
@@ -0,0 +1,277 @@
...
+    savedSegmentId: {
+      type: "string",
+      label: "Saved segment id",
+      description: "The id for an existing saved segment.",
+      optional: true,
+    },
+    prebuiltSegmentId: {
+      type: "string",
+      label: "Prebuilt segment id",
+      description: "The prebuilt segment id, if a prebuilt segment has been designated for this campaign.",
```
Comment (lcaresia) [inline components/mailchimp/actions/update-campaign/update-campaign.mjs:33]:
```suggestion
      description: "The unique ID for the campaign",
    },
    listId: {
      type: "string",
      label: "List Id",
      description: "The unique list ID",
      optional: true,
    },
    savedSegmentId: {
      type: "string",
      label: "Saved segment ID",
      description: "The ID for an existing saved segment.",
      optional: true,
    },
    prebuiltSegmentId: {
      type: "string",
      label: "Prebuilt segment ID",
      description: "The prebuilt segment ID, if a prebuilt segment has been designated for this campaign.",
```

File: components/mailchimp/actions/update-campaign/update-campaign.mjs [hunk 48]
```diff
@@ -0,0 +1,277 @@
...
+    autoTweet: {
+      type: "boolean",
+      label: "Auto tweet",
+      description: "Automatically tweet a link to the campaign archive page when the campaign is sent.",
+      optional: true,
+    },
+    autoFbPost: {
+      type: "string[]",
+      label: "Auto facebook post",
+      description: "An array of Facebook page ids to auto-post to.",
```
Comment (lcaresia) [inline components/mailchimp/actions/update-campaign/update-campaign.mjs:124]:
```suggestion
      description: "An array of Facebook page ID's to auto-post to.",
```

File: components/mailchimp/actions/update-campaign/update-campaign.mjs [hunk 49]
```diff
@@ -0,0 +1,277 @@
...
+    fbComments: {
+      type: "boolean",
+      label: "Facebook comment",
+      description: "Allows Facebook comments on the campaign (also force-enables the Campaign Archive toolbar). Defaults to true.",
+      optional: true,
+    },
+    templateId: {
+      type: "string",
+      label: "Template id",
+      description: "The id of the template to use.",
```
Comment (lcaresia) [inline components/mailchimp/actions/update-campaign/update-campaign.mjs:136]:
```suggestion
      label: "Template ID",
      description: "The ID of the template to use.",
```

File: components/mailchimp/actions/add-or-update-subscriber/add-or-update-subscriber.mjs [hunk 50]
```diff
@@ -1,35 +1,38 @@
...
+      label: "Email address",
       type: "string",
       description: "Email address for a subscriber. This value is required only if the email address is not already present on the list.",
     },
-    statu_if_new: {
+    statusIfNew: {
+      label: "Status if new",
       type: "string",
       description: "Subscriber's status. This value is required only if the email address is not already present on the list.",
       options: [
```
Comment (lcaresia) [inline components/mailchimp/actions/add-or-update-subscriber/add-or-update-subscriber.mjs:38]: Move this for the `constants.mjs` file.

File: components/mailchimp/actions/search-lists/search-lists.mjs [hunk 51]
```diff
@@ -0,0 +1,92 @@
...
+      description: "Returns files sorted by the specified field. Possible value: \"date_created\"",
+      default: "date_created",
+      optional: true,
+    },
+    sortDir: {
+      type: "string",
+      label: "Sort direction",
+      description: "Determines the order direction for sorted results. Possible values: ASC or DESC.",
+      optional: true,
+      options: [
```
Comment (lcaresia) [inline components/mailchimp/actions/search-lists/search-lists.mjs:54]: Move this for the `constants.mjs` file.

**Review by lcaresia — CHANGES_REQUESTED** [review]

File: components/mailchimp/actions/create-campaign/create-campaign.mjs (hunk 24 above)
Comment (bush4u) [reply components/mailchimp/actions/create-campaign/create-campaign.mjs:92]: I actually did not work on this.
You reviewed the part I worked on yesterday. I had to do refactoring for the others which you are reviewing now

**Comment by bush4u** [comment]
> Hello everyone, I have tested this PR and there're some test cases failed or needed improvement.
> 
> Please check the test report below for more information [MailChimp__2889.pdf](https://github.com/PipedreamHQ/pipedream/files/8771879/MailChimp__2889.pdf)

**All issues have been fixed with the some comments worth noting.**

- Add list throws a 403 error due to the free mailchimp subscription.
- Add or Update subscriber did not work using the approach of hashing email. The email field is meant to update the existing subscriber email. Also, email field cannot be optional.
- On create a list, full description for email type option has been added.

**Comment by vunguyenhung** [comment]
Hello everyone, I have re-tested this PR and there're some test cases needed small improvement. Then it should be ready for release! Great job boss!

Please check the test report below for more information
[MailChimp_2889_2220.pdf](https://github.com/PipedreamHQ/pipedream/files/8799397/MailChimp_2889_2220.pdf)

**Comment by vunguyenhung** [comment]
Hi everyone, all test cases are passed! Ready for release! 

Test report
[MailChimp_2889_2326.pdf](https://github.com/PipedreamHQ/pipedream/files/8807150/MailChimp_2889_2326.pdf)

**Review by lcaresia — APPROVED** [review]

**Pull request merged**